### PR TITLE
[KOGITO-1968] Allow milestones to trigger completion and tests

### PR DESCRIPTION
**Pull request by ruromero**
* Remove legacy Case related classes
* Add lambda to trigger event in EndNodeVisitor 
* MilestoneNodes should be completed when the event is received --> No conditions supported yet [KOGITO-2168](https://issues.redhat.com/browse/KOGITO-2168)
* Small clean-up
* Expose `milestones` method in API

Signed-off-by: ruromero <rromerom@redhat.com>

### CODE REVIEW

File: api/kogito-api/src/main/java/org/kie/api/runtime/rule/RuleRuntime.java [hunk 1]
```diff
@@ -45,6 +45,13 @@
      */
     Agenda getAgenda();
 
+    /**
+     * @return true if the Runtime supports the Agenda
+     */
+    default boolean hasAgenda() {
```
Comment (evacchi) [inline api/kogito-api/src/main/java/org/kie/api/runtime/rule/RuleRuntime.java:51]: why is this needed? if we make getAgenda() return null in the dummy implementation then you can just null check `getAgenda()`. Remember we want to get rid of the dummy implementation.

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/WorkflowProcessInstanceImpl.java [hunk 2]
```diff
@@ -142,7 +144,7 @@ public void addNodeInstance(final NodeInstance nodeInstance) {
 
     @Override
     public int getLevelForNode(String uniqueID) {
-        if ("true".equalsIgnoreCase(System.getProperty("jbpm.loop.level.disabled"))) {
+        if (Boolean.parseBoolean(System.getProperty("jbpm.loop.level.disabled"))) {
```
Comment (evacchi) [inline jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/WorkflowProcessInstanceImpl.java:147]: memo: I wonder if we should keep these overrides in Kogito

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java [hunk 3]
```diff
@@ -53,10 +53,13 @@ public void internalTrigger(final NodeInstance from, String type) {
...
-                + "-" + getMilestoneNode().getUniqueId();
-        boolean isActive = ((InternalAgenda) getProcessInstance().getKnowledgeRuntime().getAgenda())
-                .isRuleActiveInRuleFlowGroup("DROOLS_SYSTEM", rule, getProcessInstance().getId());
+        boolean isActive = true;
+        if(getProcessInstance().getKnowledgeRuntime().hasAgenda()) {
+            String rule = "RuleFlow-Milestone-" + getProcessInstance().getProcessId()
+                    + "-" + getMilestoneNode().getUniqueId();
+            isActive = ((InternalAgenda) getProcessInstance().getKnowledgeRuntime().getAgenda())
+                    .isRuleActiveInRuleFlowGroup("DROOLS_SYSTEM", rule, getProcessInstance().getId());
+        }
```
Comment (evacchi) [inline jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java:62]: I can see now where rules are used, and we should probably **write a short spec for this** and see how we can reproduce it in the best way. Rule flow groups should be considered deprecated for Drools. Maybe we can write a "System RuleUnit" in the libary if these rules are static, or we could synthesize a rule unit to be used internally.

File: api/kogito-api/src/main/java/org/kie/api/runtime/rule/RuleRuntime.java (hunk 1 above)
Comment (tiagodolphine) [reply api/kogito-api/src/main/java/org/kie/api/runtime/rule/RuleRuntime.java]: good point, +1

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/WorkflowProcessInstanceImpl.java [hunk 4]
```diff
@@ -193,22 +195,15 @@ public void removeNodeInstance(final NodeInstance nodeInstance) {
...
     @Override
     public NodeInstance getNodeInstance(String nodeInstanceId, boolean recursive) {
-        for (NodeInstance nodeInstance : getNodeInstances(recursive)) {
-            if (nodeInstance.getId().equals(nodeInstanceId)) {
-                return nodeInstance;
-            }
-        }
-        return null;
+        return getNodeInstances(recursive).stream()
+                .filter(nodeInstance -> nodeInstance.getId().equals(nodeInstanceId))
```
Comment (tiagodolphine) [inline jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/WorkflowProcessInstanceImpl.java:204]: suggestion
`.filter(nodeInstance -> Objects.equals(nodeInstance.getId(), nodeInstanceId))`

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java [hunk 5]
```diff
@@ -53,10 +53,13 @@ public void internalTrigger(final NodeInstance from, String type) {
             throw new IllegalArgumentException(
                     "A MilestoneNode only accepts default incoming connections!");
         }
-        String rule = "RuleFlow-Milestone-" + getProcessInstance().getProcessId()
-                + "-" + getMilestoneNode().getUniqueId();
-        boolean isActive = ((InternalAgenda) getProcessInstance().getKnowledgeRuntime().getAgenda())
-                .isRuleActiveInRuleFlowGroup("DROOLS_SYSTEM", rule, getProcessInstance().getId());
+        boolean isActive = true;
```
Comment (tiagodolphine) [inline jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java:56]: trying to simplify this block... just sugestion:
```
Optional.ofNullable(getProcessInstance())
                .map(WorkflowProcessInstance::getKnowledgeRuntime)
                .map(InternalKnowledgeRuntime::getAgenda)                
                .filter(InternalAgenda.class::isInstance)
                .map(InternalAgenda.class::cast)
                .map(agenda -> {
                    String rule =
                            "RuleFlow-Milestone-" + getProcessInstance().getProcessId() + "-" + getMilestoneNode().getUniqueId();
                    return agenda.isRuleActiveInRuleFlowGroup("DROOLS_SYSTEM", rule, getProcessInstance().getId());
                })
                .orElse(true);
```

File: jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java [hunk 6]
```diff
@@ -163,14 +164,20 @@ public void start(String trigger, String referenceId) {
             this.status = legacyProcessInstance.getState();
         }
     }
+
+    public void complete() {
+        if(((WorkflowProcess)legacyProcessInstance.getProcess()).isDynamic() && this.status.equals(ProcessInstance.STATE_ACTIVE)) {
```
Comment (tiagodolphine) [inline jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java:169]: Are the castings safe? WorkflowProcess, WorkflowProcessInstance, Is it worth checking instanceof before casting?

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java (hunk 3 above)
Comment (ruromero) [reply jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java]: I have filed the following JIRA https://issues.redhat.com/browse/KOGITO-2168

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java (hunk 5 above)
Comment (ruromero) [reply jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java:56]: Although your suggestion is fancier I believe the existing code is easier to understand and maintain.

File: jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java (hunk 6 above)
Comment (ruromero) [reply jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java:169]: If the legacyProcessInstance.getProcess() is always returning what is generated by the `legacyProcess()` method coming from the codegen I'd say the only possibility is that it is a RuleFlowProcess which extends WorkflowProcessImpl

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java (hunk 5 above)
Comment (tiagodolphine) [reply jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java:56]: Sure, no problem, go with it, as I said just a suggestion a matter of taste, I really like and think a more functional approach is cleaner :)

File: jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java (hunk 6 above)
Comment (tiagodolphine) [reply jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java:169]: :+1:

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java (hunk 5 above)
Comment (evacchi) [reply jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java:56]: either way, this is just a refactoring of the existing code (@ruromero only added the null check around it and reformatted a bit); it doesn't really matter because I am pretty sure all this will have to be nuked in favor of rule units 🤔

File: jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java [hunk 7]
```diff
@@ -163,14 +164,20 @@ public void start(String trigger, String referenceId) {
...
+            ((WorkflowProcessInstance)legacyProcessInstance).setState(STATE_ACTIVE);
+            this.status = ProcessInstance.STATE_COMPLETED;
```
Comment (MarianMacik) [inline jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java:171]: So Kogito process instance is completed, but legacy process instance is still active? Is that right? If so, why is it like that? I understand that cases are basically never complete as we discussed it on Zulip, but this can be considered as an inconsistency - one status is completed, the second one is active.

Comment (MarianMacik) [reply jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java:171]: Also btw - this feature is already present in 7.x? I think that we can there only close or destroy a case instance, right? We cannot explicitly complete it...

Comment (ruromero) [reply jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java:171]: Ad-hoc processes, i.e. cases, can't autocomplete unless there is a terminate end node but this is not common in cases where knowledge workers might want to add/update information in the case or trigger discretionary tasks. That means there wasn't a mechanism to close a case.

Regarding the naming. In 7.x you could close a case but here we are not explicitly using a Case model naming. We're still talking about processes and processes from the ACTIVE state move to the COMPLETED state, thus this action. Do you agree?

Comment (MarianMacik) [reply jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java:171]: I agree, but somewhat I missed the explanation why the legacy process is still being kept ACTIVE :)

Comment (ruromero) [reply jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java:171]: I get your point and that had to be aligned indeed. Thanks for your thorough review.

**Comment by ruromero** [comment]
I did a rebase against master

**Comment by tiagodolphine** [comment]
@cristianonicolai missing your review, would you like to take look?

File: api/kogito-api/src/main/java/org/kie/kogito/process/ProcessInstance.java [hunk 8]
```diff
@@ -65,7 +65,12 @@
      * @param referenceId optional reference id that points to a another  component triggering this instance
      */
     void startFrom(String nodeId, String referenceId);
-    
+
+    /**
+     * Completes a dynamic process that is in the ACTIVE state.
+     */
+    void complete();
```
Comment (cristianonicolai) [inline api/kogito-api/src/main/java/org/kie/kogito/process/ProcessInstance.java:72]: from a perspective that processes will have at least some degree of structure, would it still make sense to allow it to be completed at any time? Thinking if we should first try to go for having a minial path that demonstrates how the process will be completed, not so much like a case close action. @ruromero @krisv wdyt?

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java [hunk 9]
```diff
@@ -53,10 +53,14 @@ public void internalTrigger(final NodeInstance from, String type) {
...
+        // KOGITO-2168 Conditions not supported
+        if(getProcessInstance().getKnowledgeRuntime().getAgenda() != null) {
```
Comment (cristianonicolai) [inline jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java:58]: @ruromero is there any point for us to add this if we wont actually use this approach with the internal agenda? I guess we should simply set active to true.

File: kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+import org.kie.kogito.codegen.AbstractCodegenTest;
+import org.kie.kogito.process.Process;
+import org.kie.kogito.process.ProcessInstance;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class MilestoneTest extends AbstractCodegenTest {
+
+    @Test
+    void testSimpleMilestone() throws Exception {
```
Comment (cristianonicolai) [inline kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java:33]: @ruromero should the actual milestone state be checked here?

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java (hunk 9 above)
Comment (ruromero) [reply jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java:58]: I didn't want to remove the old functionality because other tests still rely on the agenda. Until we don't provide a replacement for conditions I prefer to keep it this way.

File: api/kogito-api/src/main/java/org/kie/kogito/process/ProcessInstance.java (hunk 8 above)
Comment (ruromero) [reply api/kogito-api/src/main/java/org/kie/kogito/process/ProcessInstance.java:72]: I think this is something we need at this point and we can adapt later on. Dynamic processes need a way to be completed.

Comment (cristianonicolai) [reply api/kogito-api/src/main/java/org/kie/kogito/process/ProcessInstance.java:72]: @ruromero sure, in the example you created the test can be completed once the milestone is achieved, we dont necessarly want to manually have to call complete in a process. We're trying to give some flexibility but not complete freedom, to me the process needs to to handle a expected path to end, otherwise people can still abort the process instance.

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java (hunk 9 above)
Comment (cristianonicolai) [reply jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java:58]: so maybe we should remove the old tests as well :) My concern is to bring things that we wont/dont need, removing later is way more complicated to assess.

File: kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java (hunk 10 above)
Comment (ruromero) [reply kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java:39]: Check my latest changes

File: api/kogito-api/src/main/java/org/kie/kogito/process/ProcessInstance.java (hunk 8 above)
Comment (ruromero) [reply api/kogito-api/src/main/java/org/kie/kogito/process/ProcessInstance.java:72]: Complete only affects dynamic processes that are active. How do you want to limit it's usage? By only allow processes without any active node to be completed for example?

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java (hunk 9 above)
Comment (ruromero) [reply jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java:58]: I'm always up for removing stuff. I just didn't dare to do it without a second opinion.

File: kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        processInstance.complete();
+        assertState(processInstance, ProcessInstance.STATE_PENDING);
+
+        processInstance.start();
+        assertState(processInstance, ProcessInstance.STATE_ACTIVE);
+        
+        RuleFlowProcessInstance legacyProcessInstance = (RuleFlowProcessInstance)((AbstractProcessInstance<?>) processInstance).legacyProcessInstance;
+        assertThat(legacyProcessInstance.getNodeInstances()).isEmpty();
+        assertThat(legacyProcessInstance.getNodeIdInError()).isNullOrEmpty();
+        Optional<String> milestoneId = Stream.of(legacyProcessInstance.getNodeContainer().getNodes()).filter(node -> node.getName().equals("Milestone1")).map(n -> (String)n.getMetaData().get(Metadata.UNIQUE_ID)).findFirst();
```
Comment (cristianonicolai) [inline kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java:61]: @ruromero I guess the idea would be to add a method into _RuleFlowProcessInstance_ with a list of miliestones, that then we could check using a status like _MilestoneStatus_ from v7. No need to publish it in the ProcessInstance object for now.

File: kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        assertThat(app).isNotNull();
+
+        Process<? extends Model> p = app.processes().processById("TestCase.SimpleMilestone");
+
+        Model m = p.createModel();
+        Map<String, Object> parameters = new HashMap<>();
+        m.fromMap(parameters);
+        ProcessInstance<?> processInstance = p.createInstance(m);
+
+        assertState(processInstance, ProcessInstance.STATE_PENDING);
```
[the remark this answers is not in the document]
Comment (ruromero) [reply kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java:51]: I want to make sure the state in the processInstance and in the legacyProcessInstance are in sync

File: kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+        Application app = generateCodeProcessesOnly("cases/SimpleMilestone.bpmn");
+        assertThat(app).isNotNull();
+
+        Process<? extends Model> p = app.processes().processById("TestCase.SimpleMilestone");
+
+        Model m = p.createModel();
+        Map<String, Object> parameters = new HashMap<>();
+        m.fromMap(parameters);
+        ProcessInstance<?> processInstance = p.createInstance(m);
```
[the remark this answers is not in the document]
Comment (ruromero) [reply kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java:49]: Not exactly this but I see your point

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java (hunk 9 above)
Comment (cristianonicolai) [reply jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/MilestoneNodeInstance.java:58]: sorry, wrong gitlog :P looks fine as is

File: kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java (hunk 11 above)
Comment (ruromero) [reply kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java]: I have added the `milestones` method as suggested

File: kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+        assertState(processInstance, ProcessInstance.STATE_PENDING);
+        processInstance.complete();
```
[the remark this answers is not in the document]
Comment (ruromero) [reply kogito-codegen/src/test/java/org/kie/kogito/process/impl/MilestoneTest.java]: As discussed I have removed the `complete` method. A terminate end node must be present in order for the dynamic process to complete

File: api/kogito-api/src/main/java/org/kie/kogito/process/CaseNodeInstance.java [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.kie.kogito.process;
+
+import java.io.Serializable;
+
+public class CaseNodeInstance implements Serializable {
```
Comment (cristianonicolai) [inline api/kogito-api/src/main/java/org/kie/kogito/process/CaseNodeInstance.java:21]:
```suggestion
public class MilestoneInstance implements Serializable {
```

File: jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java [hunk 16]
```diff
@@ -368,6 +378,35 @@ public void transitionWorkItem(String id, Transition<?> transition) {
...
+    public Collection<Milestone> milestones() {
+        return getCaseNodeInstances(MilestoneNode.class)
+                .map(n -> {
+                    String uid = (String) n.getMetaData().get(UNIQUE_ID);
+                    return new Milestone(uid, n.getName(), getStatus(uid));
+                })
+                .collect(Collectors.toList());
+    }
+
+    private Stream<Node> getCaseNodeInstances(Class<? extends Node> nodeClass) {
```
Comment (cristianonicolai) [inline jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java:391]:
```suggestion
    private Stream<Node> getNodeInstances(Class<? extends Node> nodeClass) {
```

Comment (ruromero) [reply jbpm/jbpm-flow/src/main/java/org/kie/kogito/process/impl/AbstractProcessInstance.java]: They're just nodes so I renamed it to `getNodes`
